deeprankcore/query.py

```diff
@@ -117,6 +125,130 @@ def __repr__(self) -> str:
         return f"{type(self)}({self.get_query_id()})"
 
 
+class QueryCollection:
+    """
+    Represents the collection of data queries. Queries can be saved as a dictionary to easily navigate through their data 
+    
+    """
+
+    def __init__(self):
+        self._queries = []
+
+    def add(self, query: Query):
+        """ Adds new query to the collection of all generated queries.
+            Args:
+                query (Query): must be a Query object, either ProteinProteinInterfaceResidueQuery or SingleResidueVariantAtomicQuery.
+        """
+        self._queries.append(query)
+
+    def export_dict(self, dataset_path: str):
+        """ Exports the colection of all queries to a dictionary file
+            Args:
+                dataset_path (str): the new path where the list of queries be saved to.
+        """
+        with open(dataset_path, "wb") as pkl_file:
+            pickle.dump(self, pkl_file)    
+            
+    @property
+    def queries(self) -> List[Query]:
+        return self._queries
+
+    def __contains__(self, query: Query) -> bool:
+        return query in self._queries
+
+    def __iter__(self) -> Iterator[Query]:
+        return iter(self._queries)
+
+    def _process_one_query(
+        self,
+        prefix: str,
+        feature_names: List[str],
+        query: Query):
+
+        _log.info(f'\nProcess query with process ID {os.getpid()}.')
+
+        # because only one process may access an hdf5 file at the time:
+        output_path = f"{prefix}-{os.getpid()}.hdf5"
+
+        feature_modules = [
+            importlib.import_module('deeprankcore.features.' + name) for name in feature_names]
+
+        graph = query.build(feature_modules)
+
+        graph.write_to_hdf5(output_path)
+
+    def process( # pylint: disable=too-many-locals
+        self, 
+        prefix: Optional[str] = None,
+        process_count: Optional[int] = None,
+        combine_files: bool = True,
+        feature_modules: List[ModuleType] = None
+        ) -> List[str]:
```

Comment:
Regarding the order, I agree that queries and prefix should be first. But I think it's more likely for a user to want to set the modules than to change the process count or file combining.   
So maybe:
```
def preprocess(
    queries: List[Query],
    prefix: Optional[str] = None,
    feature_modules: Union[List[ModuleType], str] = "all"
    process_count: Optional[int] = None,
    combine_files: bool = True,
    ):
```

Also, maybe rename `combine_files` to `combine_output`

Reply:
queries is no more a parameter (it's a property of the QueryCollection class now, I merged the other PR to this one). I modified the order of the former preprocess function, now process method, and the naming, like this:
```
    def process(
        self, 
        prefix: Optional[str] = None,
        feature_modules: List[ModuleType] = None,
        subprocesses: Optional[int] = None,
        combine_output: bool = True,
        ) -> List[str]:
```

Reply:
I think `processes` as an argument to `process` is slightly confusing. That's why I suggested `subprocesses` above. If you don't like that, maybe `cpus` or `cpu_count` or something like that is better?